server/db/mongodb/adapter.go

```diff
@@ -0,0 +1,2163 @@
+				return err
+			}
+
+			// Delete credentials.
+			if err = a.CredDel(uid, "", ""); err != nil {
+				return err
+			}
+
+			// And finally delete the user.
+			if _, err = a.db.Collection("users").DeleteOne(a.ctx, b.M{"_id": uid.String()}); err != nil {
+				return err
+			}
+		} else {
+			// Disable user's subscriptions.
+			if err = a.SubsDelForUser(uid, false); err != nil {
+				return err
+			}
+
+			// Disable subscriptions for topics where the user is the owner.
+			// Disable topics where the user is the owner.
+			now := t.TimeNow()
+			disable := b.M{"$set": b.M{"deletedat": now, "updatedat": now}}
+
+			if _, err = a.db.Collection("subscriptions").UpdateMany(a.ctx, topicFilter, disable); err != nil {
+				return err
+			}
+			if _, err = a.db.Collection("topics").UpdateMany(a.ctx, b.M{"_id": b.M{"$in": topicIds}}, disable); err != nil {
+				return err
+			}
+			if _, err = a.db.Collection("users").UpdateMany(a.ctx, b.M{"_id": uid.String()}, disable); err != nil {
+				return err
+			}
+		}
+
+		// Finally commit all changes
+		return sess.CommitTransaction(a.ctx)
+	}); err != nil {
+		return err
+	}
+	sess.EndSession(a.ctx)
+
+	return err
+}
+
+// UserGetDisabled returns IDs of users which were soft-deleted since given time.
+func (a *adapter) UserGetDisabled(since time.Time) ([]t.Uid, error) {
+	filter := b.M{"deletedat": b.M{"$gte": since}}
+	findOpts := mdbopts.FindOptions{Projection: b.M{"_id": 1}}
+	cur, err := a.db.Collection("users").Find(a.ctx, filter, &findOpts)
+	if err != nil {
+		return nil, err
+	}
+	defer cur.Close(a.ctx)
+
+	var uids []t.Uid
+	var userId map[string]string
+	for cur.Next(a.ctx) {
+		if err := cur.Decode(&userId); err != nil {
+			return nil, err
+		}
+		uids = append(uids, t.ParseUid(userId["_id"]))
+	}
+	return uids, nil
+}
+
+// UserUpdate updates user record
+func (a *adapter) UserUpdate(uid t.Uid, update map[string]interface{}) error {
+	// to get round the hardcoded "UpdatedAt" key in store.Users.Update()
+	update = normalizeUpdateMap(update)
+
+	_, err := a.db.Collection("users").UpdateOne(a.ctx, b.M{"_id": uid.String()}, b.M{"$set": update})
+	return err
+}
+
+// UserUpdateTags adds, removes, or resets user's tags
+func (a *adapter) UserUpdateTags(uid t.Uid, add, remove, reset []string) ([]string, error) {
+	// Compare to nil vs checking for zero length: zero length reset is valid.
+	if reset != nil {
+		// Replace Tags with the new value
+		return reset, a.UserUpdate(uid, map[string]interface{}{"tags": reset})
+	}
+
+	var user t.User
+	err := a.db.Collection("users").FindOne(a.ctx, b.M{"_id": uid.String()}).Decode(&user)
+	if err != nil {
+		return nil, err
+	}
+
+	// Mutate the tag list.
+	newTags := user.Tags
+	if len(add) > 0 {
+		newTags = union(newTags, add)
+	}
+	if len(remove) > 0 {
+		newTags = diff(newTags, remove)
+	}
+
+	update := map[string]interface{}{"tags": newTags}
+	if err := a.UserUpdate(uid, update); err != nil {
+		return nil, err
+	}
+
+	// Get the new tags
```

Comment:
Why do you need to fetch tags here?
Can you just return newTags?

Reply:
`newTags` is really not new tags, but rather a mutation operation of old tags into new tags. So, the tags must be fetched either before or after the mutation. Fetching them after the mutation seems to be safer.

Reply:
Ah, sure. Makes sense.
Can you add a comment explaining this please?